mesonbuild/compilers/swift.py

```diff
@@ -55,6 +60,22 @@ def get_werror_args(self) -> T.List[str]:
     def get_dependency_gen_args(self, outtarget: str, outfile: str) -> T.List[str]:
         return ['-emit-dependencies']
 
+    def get_dependency_compile_args(self, dep: Dependency) -> T.List[str]:
+        args = dep.get_compile_args()
+        # Some deps might sneak in a hardcoded path to an older macOS SDK, which can
+        # cause compilation errors. Let's replace all .sdk paths with the current one.
+        # SwiftPM does it this way: https://github.com/swiftlang/swift-package-manager/pull/6772
+        # Not tested on anything else than macOS for now.
+        if not self.info.is_darwin():
+            return args
+        suffix = '.sdk'
+        for idx, arg in enumerate(args):
+            if suffix not in arg:
+                continue
+            sdk_end = arg.index(suffix) + len(suffix)
+            args[idx] = arg[:2] + self.sdk_path + arg[sdk_end:]
```

Comment:
Added a guard to only affect `-I` args. Linking args are handled separately in other parts of code and I didn't want to change them because I don't have a known scenario where linking would fail, so I can't add a test either.

As for the regex, perhaps other SDKs can appear here too, like the iPhoneOS one etc.? Maybe that's why their is so permissive?

Reply:
Then we should match all those explicitly, because right now it's matching anything with `.sdk` in the path.